### PR TITLE
feat: allow message and timer boundary event migration in UI

### PR DESCRIPTION
**Pull request by pedesen**
## Description

- add hasType and hasEventType helper functions to identify element types
- add reusable element migration check
- allow message and timer boundary event migration in UI
- add/extend unit tests
- extend/add E2E test 

>  Optional: The edge cases limitations ([set 1](https://github.com/camunda/camunda/issues/19640), [set 2](https://github.com/camunda/camunda/issues/20906), [set 3](https://github.com/camunda/camunda/issues/19956)) are validated and not allowed in the frontend

These edge cases will be discussed later. 

This is the first time I tried to attach a bpmn diagram diff (created with the web modeler). It was a bit of effort, but I think it might be beneficial to understand what changed. If we see value in this I will continue to do it.

I also attached some UI screenshots to explain how the code change impacts the UI. Same here: let's keep doing it if we see value.

## Checklist

- [ ] for CI changes:
  - [ ] structural/foundational changes signed off by [CI DRI](https://github.com/cmur2)
  - [ ] [ci.yml](https://github.com/camunda/camunda/blob/main/.github/workflows/ci.yml) modifications comply with ["Unified CI" requirements](https://github.com/camunda/camunda/wiki/CI-&-Automation#workflow-inclusion-criteria)

## Related issues

related to #19886

### CODE REVIEW

File: operate/client/e2e-playwright/tests/resources/ProcessInstanceMigration/orderProcessMigration_v_2.bpmn [hunk 1]
```diff
@@ -1,20 +1,20 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<bpmn:definitions xmlns:bpmn="http://www.omg.org/spec/BPMN/20100524/MODEL" xmlns:bpmndi="http://www.omg.org/spec/BPMN/20100524/DI" xmlns:di="http://www.omg.org/spec/DD/20100524/DI" xmlns:dc="http://www.omg.org/spec/DD/20100524/DC" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:zeebe="http://camunda.org/schema/zeebe/1.0" id="Definitions_1" targetNamespace="http://bpmn.io/schema/bpmn" exporter="Camunda Modeler" exporterVersion="5.12.0">
```
Comment (pedesen) [inline operate/client/e2e-playwright/tests/resources/ProcessInstanceMigration/orderProcessMigration_v_2.bpmn:2]: Change diff:

image

File: operate/client/e2e-playwright/tests/resources/ProcessInstanceMigration/orderProcessMigration_v_3.bpmn [hunk 2]
```diff
@@ -1,20 +1,20 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<bpmn:definitions xmlns:bpmn="http://www.omg.org/spec/BPMN/20100524/MODEL" xmlns:bpmndi="http://www.omg.org/spec/BPMN/20100524/DI" xmlns:di="http://www.omg.org/spec/DD/20100524/DI" xmlns:dc="http://www.omg.org/spec/DD/20100524/DC" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:zeebe="http://camunda.org/schema/zeebe/1.0" id="Definitions_1" targetNamespace="http://bpmn.io/schema/bpmn" exporter="Camunda Modeler" exporterVersion="5.22.0">
```
Comment (pedesen) [inline operate/client/e2e-playwright/tests/resources/ProcessInstanceMigration/orderProcessMigration_v_3.bpmn:2]: Change diff:

image

File: operate/client/src/modules/mocks/diagrams/instanceMigrationBoundaryEvents.bpmn [hunk 3]
```diff
@@ -0,0 +1,270 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<bpmn:definitions xmlns:bpmn="http://www.omg.org/spec/BPMN/20100524/MODEL" xmlns:bpmndi="http://www.omg.org/spec/BPMN/20100524/DI" xmlns:dc="http://www.omg.org/spec/DD/20100524/DC" xmlns:zeebe="http://camunda.org/schema/zeebe/1.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:di="http://www.omg.org/spec/DD/20100524/DI" xmlns:modeler="http://camunda.org/schema/modeler/1.0" id="Definitions_0be8vjf" targetNamespace="http://bpmn.io/schema/bpmn" exporter="Camunda Modeler" exporterVersion="5.26.0" modeler:executionPlatform="Camunda Cloud" modeler:executionPlatformVersion="8.5.0">
```
Comment (pedesen) [inline operate/client/src/modules/mocks/diagrams/instanceMigrationBoundaryEvents.bpmn:2]: image

File: operate/client/src/modules/mocks/diagrams/instanceMigration.bpmn [hunk 4]
```diff
@@ -1,5 +1,5 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<bpmn:definitions xmlns:bpmn="http://www.omg.org/spec/BPMN/20100524/MODEL" xmlns:bpmndi="http://www.omg.org/spec/BPMN/20100524/DI" xmlns:di="http://www.omg.org/spec/DD/20100524/DI" xmlns:dc="http://www.omg.org/spec/DD/20100524/DC" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:zeebe="http://camunda.org/schema/zeebe/1.0" id="Definitions_1" targetNamespace="http://bpmn.io/schema/bpmn" exporter="Camunda Modeler" exporterVersion="5.23.0">
+<bpmn:definitions xmlns:bpmn="http://www.omg.org/spec/BPMN/20100524/MODEL" xmlns:bpmndi="http://www.omg.org/spec/BPMN/20100524/DI" xmlns:di="http://www.omg.org/spec/DD/20100524/DI" xmlns:dc="http://www.omg.org/spec/DD/20100524/DC" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:zeebe="http://camunda.org/schema/zeebe/1.0" id="Definitions_1" targetNamespace="http://bpmn.io/schema/bpmn" exporter="Camunda Modeler" exporterVersion="5.26.0">
```
Comment (pedesen) [inline operate/client/src/modules/mocks/diagrams/instanceMigration.bpmn:2]: diff:

image

File: operate/client/src/modules/mocks/diagrams/instanceMigration_v2.bpmn [hunk 5]
```diff
@@ -1,5 +1,5 @@
 <?xml version="1.0" encoding="UTF-8"?>
-<bpmn:definitions xmlns:bpmn="http://www.omg.org/spec/BPMN/20100524/MODEL" xmlns:bpmndi="http://www.omg.org/spec/BPMN/20100524/DI" xmlns:di="http://www.omg.org/spec/DD/20100524/DI" xmlns:dc="http://www.omg.org/spec/DD/20100524/DC" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:zeebe="http://camunda.org/schema/zeebe/1.0" id="Definitions_1" targetNamespace="http://bpmn.io/schema/bpmn" exporter="Camunda Modeler" exporterVersion="5.22.0">
+<bpmn:definitions xmlns:bpmn="http://www.omg.org/spec/BPMN/20100524/MODEL" xmlns:bpmndi="http://www.omg.org/spec/BPMN/20100524/DI" xmlns:di="http://www.omg.org/spec/DD/20100524/DI" xmlns:dc="http://www.omg.org/spec/DD/20100524/DC" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xmlns:zeebe="http://camunda.org/schema/zeebe/1.0" id="Definitions_1" targetNamespace="http://bpmn.io/schema/bpmn" exporter="Camunda Modeler" exporterVersion="5.26.0">
```
Comment (pedesen) [inline operate/client/src/modules/mocks/diagrams/instanceMigration_v2.bpmn:2]: diff:

image

File: operate/client/e2e-playwright/tests/processInstanceMigration.spec.ts [hunk 6]
```diff
@@ -80,7 +80,8 @@ test.describe.serial('Process Instance Migration', () => {
     await processesPage.migrationModal.confirmButton.click();
 
     // Expect auto mapping for each flow node
-    await expect(page.getByLabel(/target flow node for/i)).toHaveCount(3);
+    await expect(page.getByLabel(/target flow node for/i)).toHaveCount(11);
```
Comment (pedesen) [inline operate/client/e2e-playwright/tests/processInstanceMigration.spec.ts:83]: I added 4 tasks with 4 boundary events, so this number increased by 8.

File: operate/client/src/App/Processes/MigrationView/BottomPanel/index.tsx [hunk 7]
```diff
@@ -125,6 +127,23 @@ const BottomPanel: React.FC = observer(() => {
...
+                    if (
+                      hasType({
+                        businessObject: sourceFlowNode,
+                        types: ['bpmn:BoundaryEvent'],
+                      })
+                    ) {
+                      return (
+                        getEventType(sourceFlowNode) === getEventType(flowNode)
+                      );
+                    }
```
Comment (pedesen) [inline operate/client/src/App/Processes/MigrationView/BottomPanel/index.tsx:142]: This ensures that I can only map message events to message events and timer events to timer events in the instance migration view:

image

image

File: operate/client/src/modules/stores/processXml/utils/isMigratableFlowNode.ts [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+      businessObject,
+      types: ['bpmn:BoundaryEvent'],
+    }) &&
+    hasEventType({
+      businessObject,
+      types: ['bpmn:MessageEventDefinition', 'bpmn:TimerEventDefinition'],
+    })
+  ) {
+    return true;
+  }
```
Comment (pedesen) [inline operate/client/src/modules/stores/processXml/utils/isMigratableFlowNode.ts:28]: This ensures that boundary events are selectable in the diagram and show up in the bottom panel of instance migration mode:

image

**Review by vitorwtavares — APPROVED** [review]
Looks good, I like the `isMigratableFlowNode` solution too, nice work

**Comment by pedesen** [comment]
@Szik Please test using this image: `registry.camunda.cloud/team-operate/camunda-operate:branch-fe-19886-migrate-messate-timer-boundary`

**Review by Szik — APPROVED** [review]
